src/index.ts

```diff
@@ -4,90 +4,109 @@
+  if (isNonArrayObject(obj1) && isNonArrayObject(obj2)) {
+    return isEqualObject(obj1, obj2, ignoredObjectProps, debug);
+  }
+  return false;
+}
 
-  if (isObject(obj1) && isObject(obj2)) {
-    const keys1 = keys(obj1, ignoredObjectProps);
-    const keys2 = keys(obj2, ignoredObjectProps);
+function isEqualObject(obj1: Indexable, obj2: Indexable, ignoredObjectProps: string[], debug?: DebugFn) {
+  const ignoredProps = stringArrayToObject(ignoredObjectProps);
+  const keys1 = nonFunctionKeys(obj1);
+  const keys2 = nonFunctionKeys(obj2);
 
-    if (keys1.length != keys2.length) return false;
+  if (keys1.length != keys2.length) {
+    return false;
+  }
 
-    for (let i = 0, n = keys1.length; i < n; i++) {
-      const prop = keys1[i];
-      const val1 = obj1[prop];
-      const val2 = obj2[prop];
+  // https://stackoverflow.com/questions/17989270/for-loop-performance-storing-array-length-in-a-variable
+  const keysLength = keys1.length;
+  for (let i = 0; i < keysLength; i++) {
+    const prop = keys1[i];
+    const val1 = obj1[prop];
+    const val2 = obj2[prop];
 
-      if (!equal(val1, val2)) {
-        if (debug) debug('"%s" not equal %s:%s', prop, val1, val2);
-        return false;
+    if (!ignoredProps[prop] && !equal(val1, val2)) {
+      if (debug) {
+        debug('"%s" not equal %s:%s', prop, val1, val2);
       }
+      return false;
     }
-    return true;
   }
-  return false;
+  return true;
 }
 
 /**
  * Determine if 'val1' and 'val2' are equal
  */
 function equal(val1: any, val2: any): boolean {
-  // Special handling for NaN
-  if (typeof val1 == 'number' && isNaN(val1) && typeof val2 == 'number' && isNaN(val2)) {
+  if (isEqualNaN(val1, val2)) {
     return true;
+  } else if (Array.isArray(val1) && Array.isArray(val2)) {
+    return isEqualArray(val1, val2);
+  } else {
+    return val1 === val2 || (val1 == null && val2 == null);
   }
-
-  return (
-    val1 === val2 ||
-    // Handle null & undefined
-    (val1 == null && val2 == null) ||
-    isEqualArray(val1, val2)
-  );
 }
 
 /**
  * Determine if 'obj' is an object
  */
-function isObject(obj: any): boolean {
-  const type = typeof obj;
-
-  return 'object' == type && !Array.isArray(obj);
+function isNonArrayObject(obj: any): obj is Indexable {
+  return typeof obj === 'object' && !Array.isArray(obj);
 }
 
 /**
- * Retrieve non-ignored keys of 'obj'
+ * Retrieve non-ignored nonFunctionKeys of 'obj'
  */
-function keys(obj: { [key: string]: any }, ignoredObjectProps: Array<string>) {
-  return Object.keys(obj).filter(key => {
-    // Ignore functions
-    return 'function' != typeof obj[key] && !~ignoredObjectProps.indexOf(key);
-  });
+function nonFunctionKeys(obj: Indexable) {
+  return Object.keys(obj).filter(key => typeof obj[key] !== 'function');
 }
 
 /**
  * Determine if arrays 'arr1' and 'arr2' are equal
  */
-function isEqualArray(arr1: any, arr2: any): boolean {
-  if (Array.isArray(arr1) && Array.isArray(arr2)) {
```

Comment:
Removing this line means a `['foo']` will equal `{'0': 'foo', length: 1}` I think.